# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: security/sandbox/chromium/sandbox/win/src/interception_agent.cc
# Commit: 370c40e49230
# Full Hash: 370c40e4923073a87845b57c1f4851bacd9663da
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2019-06-12 21:42:35
# Regressor Bug: 1552160
# File Overlap Count: 2
# Description:
#   Bug 1552160 Part 1: Roll-up of chromium sandbox update and mozilla patches to get a running browser. r=jld,aklotz,tjr,bobowen
#   
#   This updates security/sandbox/chromium/ files to chromium commit 84108231f6e6e0772fb9a4643679ce76aa771e67.
#   
#   Existing and new patches applied from security/sandbox/chromium-shim/patches/with_update/ to give a compiling and mostly working browser.
# ==============================================================================

diff -r bd5ecd72f76c -r 370c40e49230 security/sandbox/chromium/sandbox/win/src/interception_agent.cc
--- a/security/sandbox/chromium/sandbox/win/src/interception_agent.cc	Wed Jun 12 21:55:59 2019 +1200
+++ b/security/sandbox/chromium/sandbox/win/src/interception_agent.cc	Wed Jun 12 11:10:48 2019 +0100
@@ -7,6 +7,8 @@
 
 #include "sandbox/win/src/interception_agent.h"
 
+#include <windows.h>
+
 #include <stddef.h>
 
 #include "sandbox/win/src/eat_resolver.h"
@@ -34,22 +36,22 @@
 SANDBOX_INTERCEPT OriginalFunctions g_originals;
 
 // Memory buffer mapped from the parent, with the list of interceptions.
-SANDBOX_INTERCEPT SharedMemory* g_interceptions = NULL;
+SANDBOX_INTERCEPT SharedMemory* g_interceptions = nullptr;
 
 InterceptionAgent* InterceptionAgent::GetInterceptionAgent() {
-  static InterceptionAgent* s_singleton = NULL;
+  static InterceptionAgent* s_singleton = nullptr;
   if (!s_singleton) {
     if (!g_interceptions)
-      return NULL;
+      return nullptr;
 
     size_t array_bytes = g_interceptions->num_intercepted_dlls * sizeof(void*);
     s_singleton = reinterpret_cast<InterceptionAgent*>(
-        new(NT_ALLOC) char[array_bytes + sizeof(InterceptionAgent)]);
+        new (NT_ALLOC) char[array_bytes + sizeof(InterceptionAgent)]);
 
     bool success = s_singleton->Init(g_interceptions);
     if (!success) {
       operator delete(s_singleton, NT_ALLOC);
-      s_singleton = NULL;
+      s_singleton = nullptr;
     }
   }
   return s_singleton;
@@ -57,8 +59,8 @@
 
 bool InterceptionAgent::Init(SharedMemory* shared_memory) {
   interceptions_ = shared_memory;
-  for (int i = 0 ; i < shared_memory->num_intercepted_dlls; i++)
-    dlls_[i] = NULL;
+  for (int i = 0; i < shared_memory->num_intercepted_dlls; i++)
+    dlls_[i] = nullptr;
   return true;
 }
 
@@ -66,8 +68,8 @@
                                  const UNICODE_STRING* name,
                                  const DllPatchInfo* dll_info) {
   UNICODE_STRING current_name;
-  current_name.Length = static_cast<USHORT>(g_nt.wcslen(dll_info->dll_name) *
-                                            sizeof(wchar_t));
+  current_name.Length =
+      static_cast<USHORT>(g_nt.wcslen(dll_info->dll_name) * sizeof(wchar_t));
   current_name.MaximumLength = current_name.Length;
   current_name.Buffer = const_cast<wchar_t*>(dll_info->dll_name);
 
@@ -93,7 +95,7 @@
       break;
 
     dll_info = reinterpret_cast<DllPatchInfo*>(
-                   reinterpret_cast<char*>(dll_info) + dll_info->record_bytes);
+        reinterpret_cast<char*>(dll_info) + dll_info->record_bytes);
   }
 
   // Return now if the dll is not in our list of interest.
@@ -111,7 +113,7 @@
   size_t buffer_bytes = offsetof(DllInterceptionData, thunks) +
                         dll_info->num_functions * sizeof(ThunkData);
   dlls_[i] = reinterpret_cast<DllInterceptionData*>(
-                 new(NT_PAGE, base_address) char[buffer_bytes]);
+      new (NT_PAGE, base_address) char[buffer_bytes]);
 
   DCHECK_NT(dlls_[i]);
   if (!dlls_[i])
@@ -137,7 +139,7 @@
   for (int i = 0; i < interceptions_->num_intercepted_dlls; i++) {
     if (dlls_[i] && dlls_[i]->base == base_address) {
       operator delete(dlls_[i], NT_PAGE);
-      dlls_[i] = NULL;
+      dlls_[i] = nullptr;
       break;
     }
   }
@@ -149,8 +151,8 @@
 // complicated.
 bool InterceptionAgent::PatchDll(const DllPatchInfo* dll_info,
                                  DllInterceptionData* thunks) {
-  DCHECK_NT(NULL != thunks);
-  DCHECK_NT(NULL != dll_info);
+  DCHECK_NT(thunks);
+  DCHECK_NT(dll_info);
 
   const FunctionInfo* function = reinterpret_cast<const FunctionInfo*>(
       reinterpret_cast<const char*>(dll_info) + dll_info->offset_to_functions);
@@ -165,8 +167,8 @@
     if (!resolver)
       return false;
 
-    const char* interceptor = function->function +
-                              g_nt.strlen(function->function) + 1;
+    const char* interceptor =
+        function->function + g_nt.strlen(function->function) + 1;
 
     if (!IsWithinRange(function, function->record_bytes, interceptor) ||
         !IsWithinRange(dll_info, dll_info->record_bytes, interceptor)) {
@@ -174,20 +176,17 @@
       return false;
     }
 
-    NTSTATUS ret = resolver->Setup(thunks->base,
-                                   interceptions_->interceptor_base,
-                                   function->function,
-                                   interceptor,
-                                   function->interceptor_address,
-                                   &thunks->thunks[i],
-                                   sizeof(ThunkData),
-                                   NULL);
+    NTSTATUS ret = resolver->Setup(
+        thunks->base, interceptions_->interceptor_base, function->function,
+        interceptor, function->interceptor_address, &thunks->thunks[i],
+        sizeof(ThunkData), nullptr);
     if (!NT_SUCCESS(ret)) {
       NOTREACHED_NT();
       return false;
     }
 
-    DCHECK_NT(!g_originals[function->id]);
+    DCHECK_NT(!g_originals[function->id] ||
+              g_originals[function->id] == &thunks->thunks[i]);
     g_originals[function->id] = &thunks->thunks[i];
 
     thunks->num_thunks++;
@@ -202,20 +201,20 @@
 
 // This method is called from within the loader lock
 ResolverThunk* InterceptionAgent::GetResolver(InterceptionType type) {
-  static EatResolverThunk* eat_resolver = NULL;
-  static SidestepResolverThunk* sidestep_resolver = NULL;
-  static SmartSidestepResolverThunk* smart_sidestep_resolver = NULL;
+  static EatResolverThunk* eat_resolver = nullptr;
+  static SidestepResolverThunk* sidestep_resolver = nullptr;
+  static SmartSidestepResolverThunk* smart_sidestep_resolver = nullptr;
 
   if (!eat_resolver)
-    eat_resolver = new(NT_ALLOC) EatResolverThunk;
+    eat_resolver = new (NT_ALLOC) EatResolverThunk;
 
 #if !defined(_WIN64)
   // Sidestep is not supported for x64.
   if (!sidestep_resolver)
-    sidestep_resolver = new(NT_ALLOC) SidestepResolverThunk;
+    sidestep_resolver = new (NT_ALLOC) SidestepResolverThunk;
 
   if (!smart_sidestep_resolver)
-    smart_sidestep_resolver = new(NT_ALLOC) SmartSidestepResolverThunk;
+    smart_sidestep_resolver = new (NT_ALLOC) SmartSidestepResolverThunk;
 #endif
 
   switch (type) {
@@ -229,7 +228,7 @@
       NOTREACHED_NT();
   }
 
-  return NULL;
+  return nullptr;
 }
 
 }  // namespace sandbox